source/sermux-proto/src/lib.rs

```diff
@@ -0,0 +1,263 @@
+    InsufficientSize,
+    /// Ran out of room while filling a buffer, this is likely
+    /// an error in the `sermux-proto` library.
+    UnexpectedBufferFull,
+    /// The cobs decoding process failed. The message was likely
+    /// malformed or not a sermux-proto frame
+    CobsDecodeFailed,
+    /// Cobs decoding succeeded, but the resulting data was not
+    /// a valid sermux-proto frame
+    MalformedFrame,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct PortChunk<'a> {
+    pub port: u16,
+    pub chunk: &'a [u8],
+}
+
+impl<'a> PortChunk<'a> {
+    /// Create a new PortChunk from the given port and data
+    #[inline]
+    pub fn new(port: u16, chunk: &'a [u8]) -> Self {
+        Self { port, chunk }
+    }
+
+    /// Calculate the size required to encode the given data payload size
+    #[inline]
+    pub fn buffer_required(&self) -> usize {
+        // Room for COBS(port:u16 + data:[u8; len]) plus a terminating zero
+        cobs::max_encoding_length(self.chunk.len() + size_of::<u16>() + 1)
+    }
+
+    /// Encodes the current [PortChunk] into the given buffer
+    pub fn encode_to<'b>(&self, out_buf: &'b mut [u8]) -> Result<&'b mut [u8], Error> {
+        let PortChunk { port, chunk } = self;
+        if out_buf.len() < self.buffer_required() {
+            return Err(Error::InsufficientSize);
+        }
+
+        let mut encoder = cobs::CobsEncoder::new(out_buf);
+        encoder
+            .push(&port.to_le_bytes())
+            .map_err(|_| Error::UnexpectedBufferFull)?;
+        encoder
+            .push(chunk)
+            .map_err(|_| Error::UnexpectedBufferFull)?;
+        let used = encoder
+            .finalize()
+            .map_err(|_| Error::UnexpectedBufferFull)?;
+        // Get the encoded amount, with room for an extra zero terminator
+        let res = out_buf
+            .get_mut(..used + 1)
+            .ok_or(Error::UnexpectedBufferFull)?;
+        res[used] = 0;
+        Ok(res)
+    }
+
+    /// Decodes a [PortChunk] from the given buffer
+    ///
+    /// NOTE: This MAY mutate `data`, even if the decoding fails.
+    pub fn decode_from(data: &'a mut [u8]) -> Result<Self, Error> {
+        let dec_len = cobs::decode_in_place(data).map_err(|_| Error::CobsDecodeFailed)?;
+
+        // Messages must have a port and at least one data byte to be
+        // well formed
+        if dec_len < (size_of::<u16>() + 1) {
+            return Err(Error::MalformedFrame);
+        }
+
+        let frame = data.get(..dec_len).ok_or(Error::MalformedFrame)?;
+
+        let mut port_bytes = [0u8; size_of::<u16>()];
+        let (port_data, chunk) = frame.split_at(size_of::<u16>());
+        port_bytes.copy_from_slice(port_data);
+        let port = u16::from_le_bytes(port_bytes);
+
+        Ok(PortChunk { port, chunk })
+    }
+
+    /// Convert into an [OwnedPortChunk]
+    ///
+    /// Only available with the `use-std` feature active
+    #[cfg(feature = "use-std")]
+    pub fn into_owned(self) -> OwnedPortChunk {
```

Comment:
nit/TIOLI: it seems like this could be an implementation of the `std::borrow::ToOwned` trait?